### DISCUSSION TITLE
Python Backports

### DISCUSSION THREAD

**Discussion by Gerry546** [meta-homeassistant/meta-homeassistant#94] [Ideas]
Now that I have used the previously proposed layer structure I still believe it is missing one element.
HomeAssistant has a tendency to use the latest (bleeding edge) variant of Python packages and it is becoming an increasing necessity to backport Python recipes directly from the master branch of Poky or meta-openembedded.

I suggest to add an additional folder called backports to easily identify them. This enables someone to check this layer, and when we catch up with the upstream layer, drop them and go back to the use of master.

**Comment by Gerry546** [comment]
Alternatively an idea could be to create a meta-homeassistant-backports layer.

**Reply by bachp** [reply]
I think a backport layer in the same repository would make sense

**Comment by moto-timo** [comment]
There is also the [meta-lts-mixins](https://git.yoctoproject.org/meta-lts-mixins/) mechanism. A Python mixin layer for 'kirkstone' is [in-progress](https://github.com/warthog9/meta-mixin-python/tree/main). There are things that could benefit the larger OE community and not just meta-homeassistant.

**Reply by Gerry546** [reply]
The meta-lts-mixins mechanism could be good, however currently at least the newest builds support the mickledore (non-lts) version of Yocto. If we (eventually) switch over to the next lts version it could be a good option.

Personally I am a bit hesitant about adding in an additional layer dependency for this layer to be able to build. Because (especially seeing what I have had to backport the last months) would mean that this additional layer would also have to be at the same speed.

However, on the other side I do agree on the fact that some recipes could indeed be submitted. What would you propose then as a strategy?

**Comment by moto-timo** [comment]
As you are probably aware, 'mickledore' is now EOL and 'nanbield' should be the supported release (with the caveat that 'nanbield' is still going through some growing pains on the Yocto Project AutoBuilder).

I have been on the fence for a long time now. My own desire to use meta-homeassistant was stalled by lack of Apple HomeKit related recipes and I know just how many dependencies that was going to bring in. On the flip side, as a maintainer of meta-python and the python infrastructure in oe-core (and as a member of the OE TSC and Board), I get frustrated by silos of goodness from which the rest of the OE community could benefit.

Ultimately, the very nature of HomeAssistant might require us to go with your meta-homeassistant-backports layer proposal. It just means we will likely continue to not be upstreaming anything to the benefit of meta-python users?

**Reply by Gerry546** [reply]
Correct, I am aware that `mickledore` is EOL and `nanbield` is the next on the list. My big problem (personally) is that I do this meta-homeassistant work as part of a home project which runs on raspberry pi products and I'm not completely convinced on their support for nanbield.

Just to clarify, I would not mind at all to upstream several recipes to meta-python. Personally it is more that I have never taken the time to go through the process that is required to upstream patches or ideas. (more like, I have no idea where to start).

But to come to your last comment, I would support to have more recipes in meta-python. But my fear is that that process is quite slow (at least compared to the HomeAssistant releases).

**Reply by moto-timo** [reply]
Well, khem actually has been doing a very good job of keeping patches from the mailing list flowing into the "master" branch (test builds on the YoE Distro using GitHub actions). We have also been working on expanding the ptest ("package test") coverage.

We don't bite. Send a patch (git send-email --to=openembedded-devel@lists.openembedded.org --subject-prefix="meta-python][PATCH" -M -1). Be patient while we walk you through the steps.
https://git.openembedded.org/meta-openembedded/tree/meta-python/README.md#n22
https://docs.yoctoproject.org/contributor-guide/index.html

**Reply by Gerry546** [reply]
Thanks, I will definitly try. But the question then is, what woudl be a good starting point. On of the recipes which I have backported from meta-python `master` to `mickledore`? I guess, since it is EOL, it will not accept new patches.

Or actually one of the newer ons like the `python3-orjson` one?

**Reply by moto-timo** [reply]
I am in the final stages of adding 'maturin' support for oe-core, which python3-orjson needs. But I haven't gotten a python3-orjson recipe to work just yet. Too many spheres being juggled.

Start with a simple recipe and submit to "master". We don't backport new recipes to stable branches (hence the mixins layer concept).

**Reply by Gerry546** [reply]
I will start trying to upstream recipes.

This does leave the question for the backports to (for instance) mickledore. I checked the `mixin` branch. But I cannot find a mixin layer which seems(?) to be widely supported. I checked here at least [openembedded-layer-index](https://layers.openembedded.org/layerindex/branch/master/layers/).

**Reply by moto-timo** [reply]
I am also currently maintaining and working on layer index. If no one submits a layer to the index, nothing happens. If a layer is broken and no one fixes it, nothing happens. If source code is left alone for 4 years, things break.

**Reply by bachp** [reply]
When I started this layer I went with the workflow of trying to upstream all changes to meta-python. However it was not easy as for some recipes home assistant needed a specific version and it got updated in meta-python. So over time more recipes accumulated here.

@moto-timo If you remember at some point we were also discussing if it would make sense to have homeassistant as just a recipe in meta-python and I wanted to keep it separate then. But maybe we should re consider the option as for me the main goal is to have the recipes distributed accross as few layers as possible. Having everything in meta-python would achive this.

**Reply by Gerry546** [reply]
Well, for now I will at least proceed with creating a local backport layer until we've concluded this discussion. That way progress on trying to follow HA itself is not hampered. :)
